### ISSUE TITLE
revisit graph annotation

### ISSUE THREAD

**Issue by ctb** [spacegraphcats/spacegraphcats#309]
The annotation approach added in #282 seems to be working out pretty well, and I had a few thoughts about how to extend and/or regularize it, as well as how to evaluate it.

My major realization is that we should focus on annotating cDBG nodes, which are invariant with respect to the catlas; these annotations can then be propagated dynamically into the catlas as we need them.

This annotation could be done in a variety of ways, including k-mers (current approach), BLAST, etc. Could also annotate taxonomically using sourmash, or sourmash + whole genomes discovered by sourmash.

In terms of measuring/evaulating things, the question is not "can we annotate cDBG nodes accurately" - that's a given, really, because it's "just" sequence annotation - but "when we annotate cDBG nodes and propagate them to the domset, is the domset annotation now accurate?" We should put some thought into how to do this for r=1, just to get started.

(Note that Taylor has mentioned results from phylophlan showing that query neighborhoods do not bring in new taxonomy.)

Also, I think it would be really interesting to explore how much of the catlas gets annotated as radius increases - this dovetails nicely with the idea of annotating cDBG nodes (and then propagating those annotations into the catlas). This could be a useful stat for people (like us!) who want to know about "unknown unknowns" that are not simply SNPs or reads that don't map.

**Comment by taylorreiter** [comment]
> My major realization is that we should focus on annotating cDBG nodes, which are invariant with respect to the catlas; these annotations can then be propagated dynamically into the catlas as we need them.

This is great. 

I have three relevant datasets. 
First, I have observations from query-by-gene, where r = 1 sometimes scoops in annotations from mostly one gene but sometimes two genes. Using the same query gene across many metagenomes, we usually get the same gene annotation in many gene neighborhoods. This is good.

Second, I have run singlem on reads from genome query nbhds. Singlem searches reads for 14 marker genes, and assigns taxonomy to a sample based on those marker genes. Some of my query genomes were contaminated so this isn't the best benchmark, but I definitely pulled in so not-the-same-taxonomy reads. So this could come from contamination, or from r=1 pulling in marker genes based on seq similarity, or heuristics in the singlem algorithm.

And the last is as you mention, strainphlan tells me that the query nbhds don't bring in new taxonomy, so yay 👍 

tldr, things probably look good but i haven't dont anything systematic!

**Comment by ctb** [comment]
re taxonomy, I just implemented a fun little script in #303, `scripts/explore-dominator-taxonomy.py`

briefly, this script -
* takes all sourmash hashes in the dataset and retrieves the cDBG unitig IDs to which they belong;
* it then aggregates those hashes under the dom nodes to which those cDBG unitigs belong;
* finally, it calculates the taxonomy of the dom node based on the LCA of the hash vals.

The goal here is to see how many dom nodes are taxonomically confused, because they touch on cDBG nodes belonging to multiple different species.

I ran it like so, on twofoo:
```
scripts/explore-dominator-taxonomy.py twofoo_k31_r1 twofoo_k31_r1_hashval_k31/index.pickle ../sourmash/podar-ref.lca.json.gz
```

and got the following results:

```
88512 dom nodes have no sourmash hashes under them.
903 dom nodes have exactly one sourmash hash under them.
37 dom nodes have two or more sourmash hashes under them.

rank of dom node lca  count of dom nodes with that rank
--------------------  ---------------------------------
strain                  35
species                  2
```

This aligns with my expectations from synthetic data - twofoo consists of reads from three genomes. Two of them are different strains from the same species, and one of them is a completely different class from the other two. What we see is that the graphs for the two different species are completely non-overlapping, at least at the resolution of sourmash k=31 scaled=1000.

More specifically, 35 of the dom nodes have cDBG nodes with two or more hashvals under them that belong to the same strain, while only two of the dom nodes contain cDBGs with hashvals from the two different strains. Of course, at the resolution we're using here, the vast majority of cDBG nodes and dom nodes have no associated hashvals, which maybe syas we should be using higher resolution sourmash or larger r! 

I'll think about exploring this on Real Metagenomes sometime, too - probably I'll look at some of taylor's HMP work.

**Comment by ctb** [comment]
Tried it out on twofoo_k31_r5 --

```
loaded 27124 nodes from catlas twofoo_k31_r5
loaded 15031 layer 1 catlas nodes
loaded 9447551 k-mers in index (0.4s)
loaded 985 hash value -> cdbg_id mappings from twofoo_k31_r5_hashval_k31/index.pickle
14149 dom nodes have no sourmash hashes under them.
796 dom nodes have exactly one sourmash hash under them.
86 dom nodes have two or more sourmash hashes under them.

rank of dom node lca  count of dom nodes with that rank
--------------------  ---------------------------------
strain                  76
species                  10
```

* higher radius 'annotates' more dom nodes with taxonomy, as expected!
* still no increase in taxonomic confusion above the species level, which is good and expected for synthetic data :)

**Comment by ctb** [comment]
if and when we find nodes that have real taxonomic confusion, it would be cool to look at what genes are in the neighborhood, too.

**Comment by taylorreiter** [comment]
Woah this is really cool. This is the first thing that I have seen that helps me develop an intuition for the distribution of species and strains with a cdbg

**Comment by ctb** [comment]
> Woah this is really cool. This is the first thing that I have seen that helps me develop an intuition for the distribution of species and strains with a cdbg

I live to serve 🙇 :)

## Post-outdoor-run thoughts --

### apply to IBD2

there's no reason we couldn't do this on your IBD2 stuff, just within the 41 genomes (or 1145, or ...all). This would presumably illuminate situations where there are overlapping portions of the graph between genomes, and/or contamination in query genomes/db.

this might be something good to do to preempt reviewer concerns about multiple "parent genomes" and/or contamination in the query genomes.

### applying to contamination analysis

you could do a variety of things to systematically examine potential MAG contaminants. one simple idea that is similar to what charcoal is currently doing with sourmash is to label all of a metagenome's cDBG nodes with the taxonomy of the constituent contigs. (you could do this with the genomes from a gather, OR a search --containment). Then, ask for cDBG nodes and/or dom nodes that have lineages where the LCA is above some threshold.

this has the advantage of bringing metagenome graph structure into the contamination analysis, with or without the use of dom nodes.

**Comment by ctb** [comment]
come to think of it, when used on a sample that maps well, e.g. something from the iHMP, this taxonomic confusion estimation approach might be a good way of evaluating what r values work well.

**Comment by taylorreiter** [comment]
I'm less concerned about contamination in this particular instance now given the query re-do (GTDB + charcoal before performing sgc queries). But I'm still very interested in graph annotation. 

I think right now I'd be particularly interested in labeling cDBG nodes with genes and taxonomy. 
With genes, my current use case would be:

1) building a catlas at r=10 using query nbhds from the same genome query performed on many metagenomes
2) performing abundance estimation using #303 
3) performing differential abundance analysis with these results, e.g. with corncob
4) visualizing the catlases, colored by more or less abundant in disease. (I think many will just be less abundant, and these aren't that interesting. If the whole graph disappears in disease, then the species basically just left the microbiome. that's pretty easy to interpret without any more analysis. But for the species where one strain becomes more abundant, only part of the catlas will become more abundant, and then **i'll want to know what those genes are**).

so...what would be really useful is being able to use a multifasta-eque approach to label nodes with genes. 
I think what I would do in this case, given that this is the gut and there are many, many genomes available for my organisms of interest, I would create a "pangenome" with e.g. [roary](https://sanger-pathogens.github.io/Roary/) that represents all of the known genes for a given species. Then query the catlas with those genes, and if there's a match, give the node the annotation? Basically a catlas no id <-> gene id csv file feels like it would be really useful here.

Is there a better idea here? I don't think hashes will give a complete picture of everything going on...although I guess i'll also have to go back to hashes eventually too and make sure downstream results match upstream results.
